### DISCUSSION TITLE
How do we know when persistent layer is loaded and/or whether values were loaded?

### DISCUSSION THREAD

**Discussion by djMax** [LegendApp/legend-state#115] [Q&A]
As we end up with many observables in an MMKV/RN app, how do we bind some sort of event to when the values are loaded from the persistence layer? Are all those loads synchronous? Is there some way to mess with `when` to know that it has loaded from the persistent layer?

**Comment by jmeistrich** [comment]
`persistObservable` returns an Observable with some properties for reading/modifying status. You can use the `isLoadedLocal` property on there to know when it's loaded:

```js
const state = persistObservable(State.demo, {
    local: 'demo'
})
await when(state.isLoadedLocal)
...
```

But MMKV is synchronous so you don't even need to worry about it, and same with LocalStorage on web which is also synchronous. The `isLoadedLocal` property is useful for the IndexedDB plugin on web and a potential future AsyncStorage plugin

**Comment by djMax** [comment]
that's useful info, thank you. However, does this mean that if we use a "lot" of legend-based persistent state, it will lock the UI thread while loading?

**Reply by jmeistrich** [reply]
I chose to start with mmkv because it seems like it's the fastest option by quite a margin. Legend has tons of data flowing through the IndexedDB plugin (and a custom Electron -> .json plugin) so I know the full system is fast. But Bravely is our only app using mmkv and it stores only a moderate amount of data so I can't say for sure.

I'd say try sending a lot of state through mmkv and see how fast it loads. If it's too slow you could split it into multiple observables that you load as needed - just call persistObservable when you're ready to load it. You can also have one large state object and persist individual subtrees in different ways if you prefer - we do that in Bravely.

If you're aware of a faster option I'd be very happy to help make a plugin for it, but AFAIK react-native-mmkv is the fastest.
